File: frontend/api/app/services/sms_parser.py
import re
import dateparser
from decimal import Decimal
from datetime import datetime, date
from typing import Optional, Dict, Any, List
from app.models import TransactionType, TransactionSource


SMS_PATTERNS = {
    "amount": [
        r'(?:rs\.?|inr|₹)\s*(\d+(?:,\d{3})*(?:\.\d{2})?)',
        r'(\d+(?:,\d{3})*(?:\.\d{2})?)\s*(?:rs\.?|inr|₹)',
        r'(?:spent|paid|debited|charged|purchase)\s+(?:rs\.?|inr|₹)?\s*(\d+(?:,\d{3})*(?:\.\d{2})?)',
        r'(?:credited|received|deposited)\s+(?:rs\.?|inr|₹)?\s*(\d+(?:,\d{3})*(?:\.\d{2})?)',
    ],
    "merchant": [
        r'(?:at|from|to|via)\s+([A-Za-z0-9\s&\-.]+?)(?:\s+(?:on|at|for|using|via|ref|txn|transaction|amount|rs|inr|₹)|$)',
        r'(?:merchant|payee):\s*([A-Za-z0-9\s&\-.]+)',
        r'(?:to|paid to)\s+([A-Za-z0-9\s&\-.]+?)(?:\s+(?:rs|inr|₹|\d)|$)',
    ],
    "date": [
        r'(?:on|dated)\s+(\d{1,2}[-/]\d{1,2}[-/]\d{2,4})',
        r'(?:on|dated)\s+(\d{1,2}\s+\w{3}\s+\d{2,4})',
    ],
    "card_last4": [
        r'(?:card|ending|xxx)\s*(\d{4})',
        r'(\d{4})\s*(?:debited|credited)',
    ]
}

CATEGORY_KEYWORDS = {
    "Food & Dining": ["swiggy", "zomato", "restaurant", "cafe", "food", "dining", "pizza", "burger", "dominos", "kfc", "mcdonalds", "subway", "starbucks", "cafe coffee day"],
    "Transport": ["uber", "ola", "rapido", "metro", "bus", "train", "fuel", "petrol", "diesel", "indian oil", "hp petrol", "bharat petroleum"],
    "Shopping": ["amazon", "flipkart", "myntra", "ajio", "shopping", "purchase", "nykaa", "meesho"],
    "Entertainment": ["netflix", "prime", "hotstar", "bookmyshow", "movie", "entertainment", "sony liv", "zee5", "spotify", "youtube premium"],
    "Bills & Utilities": ["electricity", "water", "gas", "broadband", "wifi", "mobile", "recharge", "bill", "airtel", "jio", "vi ", "bsnl", "tatapower", "bescom"],
    "Healthcare": ["hospital", "clinic", "pharmacy", "medical", "health", "doctor", "apollo", "fortis", "max hospital", "medplus", "1mg", "pharmeasy"],
    "Education": ["school", "college", "university", "course", "udemy", "coursera", "education", "byju", "unacademy"],
    "Salary": ["salary", "credited", "salary credited", "payroll"],
    "Investments": ["mutual fund", "sip", "zerodha", "groww", "upstox", "coin", "smallcase", "etmoney", "investment"],
}


def parse_sms_message(content: str, sender: str, received_at: datetime) -> Optional[Dict[str, Any]]:
    content_lower = content.lower()
    sender_lower = sender.lower()

    if not any(keyword in content_lower for keyword in ["rs", "inr", "₹", "debited", "credited", "spent", "paid", "charged", "purchase", "received", "deposited", "salary"]):
        return None

    amount = None
    for pattern in SMS_PATTERNS["amount"]:
        match = re.search(pattern, content_lower)
        if match:
            amount = Decimal(match.group(1).replace(",", ""))
            break

    if amount is None:
        return None

    merchant = None
    for pattern in SMS_PATTERNS["merchant"]:
        match = re.search(pattern, content_lower)
        if match:
            merchant = match.group(1).strip().title()
            break

    if not merchant:
        sender_keywords = ["bank", "hdfc", "icici", "sbi", "axis", "kotak", "yes bank", "idfc", "indusind", "federal", "rbl", "citi", "hsbc", "standard chartered"]
        for keyword in sender_keywords:
            if keyword in sender_lower:
                merchant = keyword.title()
                break

    trans_date = received_at.date()
    for pattern in SMS_PATTERNS["date"]:
        match = re.search(pattern, content_lower)
        if match:
            try:
                trans_date = dateparser.parse(match.group(1)).date()
                break
            except:
                pass

    txn_type = TransactionType.EXPENSE
    if any(word in content_lower for word in ["credited", "received", "deposited", "salary", "refund", "cashback", "reversal"]):
        txn_type = TransactionType.INCOME

    category = "Others"
    for cat, keywords in CATEGORY_KEYWORDS.items():
        if any(kw in content_lower for kw in keywords):
            category = cat
            break

    if merchant:
        for cat, keywords in CATEGORY_KEYWORDS.items():
            if any(kw in merchant.lower() for kw in keywords):
                category = cat
                break

    confidence = 0.8 if amount and merchant else (0.6 if amount else 0.3)

    return {
        "amount": amount,
        "currency": "INR",
        "type": txn_type,
        "merchant_name": merchant,
        "transaction_date": trans_date,
        "description": content[:500],
        "category": category,
        "confidence": confidence,
        "raw_message": content
    }


def parse_email_message(subject: str, body: str, sender: str, received_at: datetime) -> Optional[Dict[str, Any]]:
    content = f"{subject} {body}"
    return parse_sms_message(content, sender, received_at)


async def get_or_create_category(name: str, user, db) -> int:
    from app.models import Category
    from sqlalchemy import select

    result = await db.execute(select(Category).where(Category.name == name, Category.user_id == user.id))
    cat = result.scalar_one_or_none()
    if cat:
        return cat.id

    cat = Category(name=name, user_id=user.id, is_default=False)
    db.add(cat)
    await db.flush()
    return cat.id

File: android-app/app/src/main/java/com/expensetracker/app/data/parser/SmsParser.kt
package com.expensetracker.app.data.parser

import com.expensetracker.app.data.model.ParsedMessage
import com.expensetracker.app.data.model.Transaction
import com.expensetracker.app.data.model.TransactionSource
import com.expensetracker.app.data.model.TransactionType
import java.math.BigDecimal
import java.math.RoundingMode
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.util.regex.Matcher
import java.util.regex.Pattern

object SmsParser {

    private val AMOUNT_PATTERNS = listOf(
        Pattern.compile("(?i)(?:rs\\.?|inr|₹)\\s*([\\d,]+\\.?\\d*)"),
        Pattern.compile("(?i)([\\d,]+\\.?\\d*)\\s*(?:rs\\.?|inr|₹)"),
        Pattern.compile("(?i)(?:spent|paid|debited|charged|purchase)\\s+(?:rs\\.?|inr|₹)?\\s*([\\d,]+\\.?\\d*)"),
        Pattern.compile("(?i)(?:credited|received|deposited)\\s+(?:rs\\.?|inr|₹)?\\s*([\\d,]+\\.?\\d*)"),
    )

    private val BALANCE_PATTERNS = listOf(
        Pattern.compile("(?i)(?:bal|balance|avbl|avl)\\s*(?:rs\\.?|inr|₹)?\\s*([\\d,]+\\.?\\d*)"),
        Pattern.compile("(?i)Available\\s+Credit\\s+Limit\\s*(?:rs\\.?|inr|₹)?\\s*([\\d,]+\\.?\\d*)")
    )

    private val MERCHANT_PATTERNS = listOf(
        Pattern.compile("(?i)(?:at|from|to|via)\\s+([A-Za-z0-9\\s&\\-.]+?)(?:\\s+(?:on|at|for|using|via|ref|txn|transaction|amount|rs|inr|₹)|$)"),
        Pattern.compile("(?i)(?:merchant|payee):\\s*([A-Za-z0-9\\s&\\-.]+)"),
    )

    private val DATE_PATTERNS = listOf(
        Pattern.compile("(?i)(?:on|dated)\\s+(\\d{1,2}[-/]\\d{1,2}[-/]\\d{2,4})"),
        Pattern.compile("(?i)(?:on|dated)\\s+(\\d{1,2}\\s+\\w{3}\\s+\\d{2,4})"),
    )

    private val CATEGORY_KEYWORDS = mapOf(
        "Food & Dining" to setOf("swiggy", "zomato", "restaurant", "cafe", "food", "dining", "pizza", "burger", "dominos", "kfc", "mcdonalds", "subway", "starbucks", "cafe coffee day", "blinkit", "zepto", "bigbasket"),
        "Transport" to setOf("uber", "ola", "rapido", "metro", "bus", "train", "fuel", "petrol", "diesel", "indian oil", "hp petrol", "bharat petroleum"),
        "Shopping" to setOf("amazon", "flipkart", "myntra", "ajio", "shopping", "purchase", "nykaa", "meesho"),
        "Entertainment" to setOf("netflix", "prime", "hotstar", "bookmyshow", "movie", "entertainment", "sony liv", "zee5", "spotify", "youtube premium", "google play", "apple"),
        "Bills & Utilities" to setOf("electricity", "water", "gas", "broadband", "wifi", "mobile", "recharge", "bill", "airtel", "jio", "vi ", "bsnl", "tatapower", "bescom"),
        "Healthcare" to setOf("hospital", "clinic", "pharmacy", "medical", "health", "doctor", "apollo", "fortis", "max hospital", "medplus", "1mg", "pharmeasy"),
        "Education" to setOf("school", "college", "university", "course", "udemy", "coursera", "education", "byju", "unacademy"),
        "Salary" to setOf("salary", "credited", "salary credited", "payroll"),
        "Investments" to setOf("mutual fund", "sip", "zerodha", "groww", "upstox", "coin", "smallcase", "etmoney", "investment"),
    )

    fun parseSms(content: String, sender: String, receivedAt: LocalDateTime): ParsedMessage? {
        val contentLower = content.lowercase()
        val senderLower = sender.lowercase()

        val amount = extractAmount(contentLower)
        if (amount == null) return null

        val merchant = extractMerchant(contentLower)
        val date = extractDate(contentLower) ?: receivedAt.toLocalDate()
        val type = determineType(contentLower)
        val category = categorizeMerchant(merchant ?: contentLower)
        val confidence = calculateConfidence(amount != null, merchant != null)

        return ParsedMessage(
            userId = 0,
            source = TransactionSource.SMS,
            rawContent = content,
            sender = sender,
            receivedAt = receivedAt,
            parsedAmount = amount,
            parsedCurrency = "INR",
            parsedMerchant = merchant,
            parsedCategory = category,
            parsedDate = date,
            parsedType = type,
            confidenceScore = confidence
        )
    }

    private fun extractAmount(text: String): Double? {
        for (pattern in AMOUNT_PATTERNS) {
            val matcher = pattern.matcher(text)
            if (matcher.find()) {
                try {
                    return matcher.group(1).replace(",", "").toDouble()
                } catch (e: NumberFormatException) {
                    continue
                }
            }
        }
        return null
    }

    private fun extractMerchant(text: String): String? {
        for (pattern in MERCHANT_PATTERNS) {
            val matcher = pattern.matcher(text)
            if (matcher.find()) {
                val merchant = matcher.group(1).trim()
                if (merchant.length >= 2 && merchant.length <= 100) {
                    return merchant.split(" ").joinToString(" ") { it.replaceFirstChar { char -> char.uppercase() } }
                }
            }
        }
        return null
    }

    private fun extractDate(text: String): LocalDate? {
        for (pattern in DATE_PATTERNS) {
            val matcher = pattern.matcher(text)
            if (matcher.find()) {
                try {
                    val dateStr = matcher.group(1)
                    return parseDate(dateStr)
                } catch (e: Exception) {
                    continue
                }
            }
        }
        return null
    }

    private fun parseDate(dateStr: String): LocalDate? {
        val formatters = listOf(
            DateTimeFormatter.ofPattern("dd-MM-yyyy"),
            DateTimeFormatter.ofPattern("dd/MM/yyyy"),
            DateTimeFormatter.ofPattern("dd-MM-yy"),
            DateTimeFormatter.ofPattern("dd/MM/yy"),
            DateTimeFormatter.ofPattern("d MMM yyyy"),
            DateTimeFormatter.ofPattern("d MMM yy"),
        )
        for (formatter in formatters) {
            try {
                return LocalDate.parse(dateStr, formatter)
            } catch (e: Exception) {
                continue
            }
        }
        return null
    }

    private fun determineType(text: String): TransactionType {
        val creditKeywords = setOf("credited", "received", "deposited", "salary", "refund", "cashback", "reversal")
        return if (creditKeywords.any { text.contains(it) }) TransactionType.INCOME else TransactionType.EXPENSE
    }

    private fun categorizeMerchant(text: String): String {
        val textLower = text.lowercase()
        for ((category, keywords) in CATEGORY_KEYWORDS) {
            if (keywords.any { textLower.contains(it) }) {
                return category
            }
        }
        return "Others"
    }

    private fun calculateConfidence(hasAmount: Boolean, hasMerchant: Boolean): Double {
        return when {
            hasAmount && hasMerchant -> 0.9
            hasAmount -> 0.6
            else -> 0.3
        }
    }
}
